### ISSUE TITLE
Compilation failing: use of undeclared identifier 'LONG_LONG_MAX'. macOS 10.15.2

### ISSUE THREAD

**Issue by rileytwo** [r-lib/vroom#208]
Hi,

I'm trying to install `vroom` on macOS 10.15.2. I installed R from source from this tap: https://github.com/sethrfore/homebrew-r-srf.

So far, I haven't had any issues installing packages--`Rcpp`, `devtools`, `tensorflow`, etc. installed without any error. This leads me to believe that the issue isn't with my dev environment... but, I don't know enough about C++ to reason about why the installation is failing--I may be missing something obvious!

The problem with `vroom` I'm having involves [src/vroom_big_int.cc](https://github.com/r-lib/vroom/blob/master/src/vroom_big_int.cc). `'LONG_LONG_MAX'` specifically. Is there something I'm missing when trying to install `vroom`? I tried google to see if I'm missing a header file somewhere that declares `LONG_LONG_MAX`, but I wasn't able to come to any conclusions for what I may be missing.

Below I've put my `~/.R/Makevars` file and the output when trying to install `vroom` from CRAN (the same error occurs when trying to install using `remotes::install_github('r-lib/vroom')`)

<details>
<summary><code>~/.R/Makevars</code></summary>

```make
# --------
# Makevars
# vim: filetype=make
# --------

# Homebrew bin / opt / lib locations
HB=/usr/local/bin
HO=/usr/local/opt
HL=/usr/local/lib
HI=/usr/local/include

# macOS SDK
# use this if you installed Xcode CLT without Xcode
SDK=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk
#
# use this if you installed Xcode CLT with Xcode
#XCODE=/Applications/Xcode.app/Contents
#SDK=$(XCODE)/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk

# ccache
CCACHE=$(HB)/ccache

# Make using all cores (set number to your number of cores)
MAKE=make -j`nproc`

# GNU version (assumes latest in Homebrew as of May 2019)
VER=9

# LLVM (Clang) compiler options
CC=$(CCACHE) $(HO)/llvm/bin/clang
CXX=$(CC)++
CXX98=$(CC)++
CXX11=$(CC)++
CXX14=$(CC)++
CXX17=$(CC)++

# FORTRAN
FC=$(CCACHE) $(HB)/gfortran-$(VER)
F77=$(FC)
FLIBS=-L$(HL)/gcc/$(VER) -lgfortran -lquadmath -lm

# STD libraries
CXX1XSTD=-std=c++0x
CXX11STD=-std=c++11
CXX14STD=-std=c++14
CXX17STD=-std=c++17

# FLAGS
STD_FLAGS=-g -O3 -Wall -pedantic -mtune=native -pipe
CFLAGS=$(STD_FLAGS)
CXXFLAGS=$(STD_FLAGS)
CXX98FLAGS=$(STD_FLAGS)
CXX11FLAGS=$(STD_FLAGS)
CXX14FLAGS=$(STD_FLAGS)
CXX17FLAGS=$(STD_FLAGS)

# Preprocessor FLAGS
# NB: -isysroot refigures the include path to the Xcode SDK
CPPFLAGS=-isysroot $(SDK) -I$(HI) \
		 -I$(HO)/llvm/include \
		 -I$(HO)/openssl/include \
		 -I$(HO)/gettext/include \
		 -I$(HO)/tcl-tk/include \
		 -I$(HO)/jpeg/include \
		 -I$(HO)/libpng/include

LDFLAGS+=-L$(HO)/llvm/lib -Wl,-rpath,$(HO)/llvm/lib \
		 -L$(HO)/openssl/lib \
		 -L$(HO)/gettext/lib \
		 -L$(HO)/jpeg/lib \
		 -L$(HO)/libpng/lib

# R-specific flags for OpenMP support
SHLIB_OPENMP_CFLAGS=-fopenmp
SHLIB_OPENMP_CXXFLAGS=-fopenmp
SHLIB_OPENMP_CXX98FLAGS=-fopenmp
SHLIB_OPENMP_CXX11FLAGS=-fopenmp
SHLIB_OPENMP_CXX14FLAGS=-fopenmp
SHLIB_OPENMP_CXX17FLAGS=-fopenmp
SHLIB_OPENMP_FCFLAGS=-fopenmp
SHLIB_OPENMP_FFLAGS=-fopenmp
```

</details>

<details>
<summary>Installation</summary>

```r
install.packages('vroom')
trying URL 'https://cloud.r-project.org/src/contrib/vroom_1.2.0.tar.gz'
Content type 'application/x-gzip' length 729431 bytes (712 KB)
\==================================================
downloaded 712 KB

* installing *source* package ‘vroom’ ...
** package ‘vroom’ successfully unpacked and MD5 sums checked
** using staged installation
** libs
/usr/local/bin/ccache /usr/local/opt/llvm/bin/clang++ -std=c++11 -I"/usr/local/Cellar/r/3.6.2/lib/R/include" -DNDEBUG  -I"/usr/local/Cellar/r/3.6.2/lib/R/library/progress/include" -I"/usr/local/Cellar/r/3.6.2/lib/R/library/Rcpp/include" -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -I/usr/local/opt/llvm/include -I/usr/local/opt/openssl/include -I/usr/local/opt/gettext/include -I/usr/local/opt/tcl-tk/include -I/usr/local/opt/jpeg/include -I/usr/local/opt/libpng/include -Imio/include -DWIN32_LEAN_AND_MEAN -Ispdlog/include -fPIC  -g -O3 -Wall -pedantic -mtune=native -pipe -c Iconv.cpp -o Iconv.o
/usr/local/bin/ccache /usr/local/opt/llvm/bin/clang++ -std=c++11 -I"/usr/local/Cellar/r/3.6.2/lib/R/include" -DNDEBUG  -I"/usr/local/Cellar/r/3.6.2/lib/R/library/progress/include" -I"/usr/local/Cellar/r/3.6.2/lib/R/library/Rcpp/include" -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -I/usr/local/opt/llvm/include -I/usr/local/opt/openssl/include -I/usr/local/opt/gettext/include -I/usr/local/opt/tcl-tk/include -I/usr/local/opt/jpeg/include -I/usr/local/opt/libpng/include -Imio/include -DWIN32_LEAN_AND_MEAN -Ispdlog/include -fPIC  -g -O3 -Wall -pedantic -mtune=native -pipe -c LocaleInfo.cpp -o LocaleInfo.o
/usr/local/bin/ccache /usr/local/opt/llvm/bin/clang++ -std=c++11 -I"/usr/local/Cellar/r/3.6.2/lib/R/include" -DNDEBUG  -I"/usr/local/Cellar/r/3.6.2/lib/R/library/progress/include" -I"/usr/local/Cellar/r/3.6.2/lib/R/library/Rcpp/include" -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -I/usr/local/opt/llvm/include -I/usr/local/opt/openssl/include -I/usr/local/opt/gettext/include -I/usr/local/opt/tcl-tk/include -I/usr/local/opt/jpeg/include -I/usr/local/opt/libpng/include -Imio/include -DWIN32_LEAN_AND_MEAN -Ispdlog/include -fPIC  -g -O3 -Wall -pedantic -mtune=native -pipe -c RcppExports.cpp -o RcppExports.o
/usr/local/bin/ccache /usr/local/opt/llvm/bin/clang++ -std=c++11 -I"/usr/local/Cellar/r/3.6.2/lib/R/include" -DNDEBUG  -I"/usr/local/Cellar/r/3.6.2/lib/R/library/progress/include" -I"/usr/local/Cellar/r/3.6.2/lib/R/library/Rcpp/include" -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -I/usr/local/opt/llvm/include -I/usr/local/opt/openssl/include -I/usr/local/opt/gettext/include -I/usr/local/opt/tcl-tk/include -I/usr/local/opt/jpeg/include -I/usr/local/opt/libpng/include -Imio/include -DWIN32_LEAN_AND_MEAN -Ispdlog/include -fPIC  -g -O3 -Wall -pedantic -mtune=native -pipe -c altrep.cc -o altrep.o
/usr/local/bin/ccache /usr/local/opt/llvm/bin/clang++ -std=c++11 -I"/usr/local/Cellar/r/3.6.2/lib/R/include" -DNDEBUG  -I"/usr/local/Cellar/r/3.6.2/lib/R/library/progress/include" -I"/usr/local/Cellar/r/3.6.2/lib/R/library/Rcpp/include" -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -I/usr/local/opt/llvm/include -I/usr/local/opt/openssl/include -I/usr/local/opt/gettext/include -I/usr/local/opt/tcl-tk/include -I/usr/local/opt/jpeg/include -I/usr/local/opt/libpng/include -Imio/include -DWIN32_LEAN_AND_MEAN -Ispdlog/include -fPIC  -g -O3 -Wall -pedantic -mtune=native -pipe -c delimited_index.cc -o delimited_index.o
/usr/local/bin/ccache /usr/local/opt/llvm/bin/clang++ -std=c++11 -I"/usr/local/Cellar/r/3.6.2/lib/R/include" -DNDEBUG  -I"/usr/local/Cellar/r/3.6.2/lib/R/library/progress/include" -I"/usr/local/Cellar/r/3.6.2/lib/R/library/Rcpp/include" -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -I/usr/local/opt/llvm/include -I/usr/local/opt/openssl/include -I/usr/local/opt/gettext/include -I/usr/local/opt/tcl-tk/include -I/usr/local/opt/jpeg/include -I/usr/local/opt/libpng/include -Imio/include -DWIN32_LEAN_AND_MEAN -Ispdlog/include -fPIC  -g -O3 -Wall -pedantic -mtune=native -pipe -c delimited_index_connection.cc -o delimited_index_connection.o
/usr/local/bin/ccache /usr/local/opt/llvm/bin/clang++ -std=c++11 -I"/usr/local/Cellar/r/3.6.2/lib/R/include" -DNDEBUG  -I"/usr/local/Cellar/r/3.6.2/lib/R/library/progress/include" -I"/usr/local/Cellar/r/3.6.2/lib/R/library/Rcpp/include" -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -I/usr/local/opt/llvm/include -I/usr/local/opt/openssl/include -I/usr/local/opt/gettext/include -I/usr/local/opt/tcl-tk/include -I/usr/local/opt/jpeg/include -I/usr/local/opt/libpng/include -Imio/include -DWIN32_LEAN_AND_MEAN -Ispdlog/include -fPIC  -g -O3 -Wall -pedantic -mtune=native -pipe -c fixed_width_index_connection.cc -o fixed_width_index_connection.o
/usr/local/bin/ccache /usr/local/opt/llvm/bin/clang++ -std=c++11 -I"/usr/local/Cellar/r/3.6.2/lib/R/include" -DNDEBUG  -I"/usr/local/Cellar/r/3.6.2/lib/R/library/progress/include" -I"/usr/local/Cellar/r/3.6.2/lib/R/library/Rcpp/include" -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -I/usr/local/opt/llvm/include -I/usr/local/opt/openssl/include -I/usr/local/opt/gettext/include -I/usr/local/opt/tcl-tk/include -I/usr/local/opt/jpeg/include -I/usr/local/opt/libpng/include -Imio/include -DWIN32_LEAN_AND_MEAN -Ispdlog/include -fPIC  -g -O3 -Wall -pedantic -mtune=native -pipe -c gen.cc -o gen.o
/usr/local/bin/ccache /usr/local/opt/llvm/bin/clang -I"/usr/local/Cellar/r/3.6.2/lib/R/include" -DNDEBUG  -I"/usr/local/Cellar/r/3.6.2/lib/R/library/progress/include" -I"/usr/local/Cellar/r/3.6.2/lib/R/library/Rcpp/include" -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -I/usr/local/opt/llvm/include -I/usr/local/opt/openssl/include -I/usr/local/opt/gettext/include -I/usr/local/opt/tcl-tk/include -I/usr/local/opt/jpeg/include -I/usr/local/opt/libpng/include  -fPIC  -g -O3 -Wall -pedantic -mtune=native -pipe -c grisu3.c -o grisu3.o
/usr/local/bin/ccache /usr/local/opt/llvm/bin/clang++ -std=c++11 -I"/usr/local/Cellar/r/3.6.2/lib/R/include" -DNDEBUG  -I"/usr/local/Cellar/r/3.6.2/lib/R/library/progress/include" -I"/usr/local/Cellar/r/3.6.2/lib/R/library/Rcpp/include" -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -I/usr/local/opt/llvm/include -I/usr/local/opt/openssl/include -I/usr/local/opt/gettext/include -I/usr/local/opt/tcl-tk/include -I/usr/local/opt/jpeg/include -I/usr/local/opt/libpng/include -Imio/include -DWIN32_LEAN_AND_MEAN -Ispdlog/include -fPIC  -g -O3 -Wall -pedantic -mtune=native -pipe -c guess_type.cc -o guess_type.o
/usr/local/bin/ccache /usr/local/opt/llvm/bin/clang++ -std=c++11 -I"/usr/local/Cellar/r/3.6.2/lib/R/include" -DNDEBUG  -I"/usr/local/Cellar/r/3.6.2/lib/R/library/progress/include" -I"/usr/local/Cellar/r/3.6.2/lib/R/library/Rcpp/include" -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -I/usr/local/opt/llvm/include -I/usr/local/opt/openssl/include -I/usr/local/opt/gettext/include -I/usr/local/opt/tcl-tk/include -I/usr/local/opt/jpeg/include -I/usr/local/opt/libpng/include -Imio/include -DWIN32_LEAN_AND_MEAN -Ispdlog/include -fPIC  -g -O3 -Wall -pedantic -mtune=native -pipe -c index_collection.cc -o index_collection.o
/usr/local/bin/ccache /usr/local/opt/llvm/bin/clang -I"/usr/local/Cellar/r/3.6.2/lib/R/include" -DNDEBUG  -I"/usr/local/Cellar/r/3.6.2/lib/R/library/progress/include" -I"/usr/local/Cellar/r/3.6.2/lib/R/library/Rcpp/include" -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -I/usr/local/opt/llvm/include -I/usr/local/opt/openssl/include -I/usr/local/opt/gettext/include -I/usr/local/opt/tcl-tk/include -I/usr/local/opt/jpeg/include -I/usr/local/opt/libpng/include  -fPIC  -g -O3 -Wall -pedantic -mtune=native -pipe -c localtime.c -o localtime.o
/usr/local/bin/ccache /usr/local/opt/llvm/bin/clang++ -std=c++11 -I"/usr/local/Cellar/r/3.6.2/lib/R/include" -DNDEBUG  -I"/usr/local/Cellar/r/3.6.2/lib/R/library/progress/include" -I"/usr/local/Cellar/r/3.6.2/lib/R/library/Rcpp/include" -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -I/usr/local/opt/llvm/include -I/usr/local/opt/openssl/include -I/usr/local/opt/gettext/include -I/usr/local/opt/tcl-tk/include -I/usr/local/opt/jpeg/include -I/usr/local/opt/libpng/include -Imio/include -DWIN32_LEAN_AND_MEAN -Ispdlog/include -fPIC  -g -O3 -Wall -pedantic -mtune=native -pipe -c vroom.cc -o vroom.o
/usr/local/bin/ccache /usr/local/opt/llvm/bin/clang++ -std=c++11 -I"/usr/local/Cellar/r/3.6.2/lib/R/include" -DNDEBUG  -I"/usr/local/Cellar/r/3.6.2/lib/R/library/progress/include" -I"/usr/local/Cellar/r/3.6.2/lib/R/library/Rcpp/include" -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -I/usr/local/opt/llvm/include -I/usr/local/opt/openssl/include -I/usr/local/opt/gettext/include -I/usr/local/opt/tcl-tk/include -I/usr/local/opt/jpeg/include -I/usr/local/opt/libpng/include -Imio/include -DWIN32_LEAN_AND_MEAN -Ispdlog/include -fPIC  -g -O3 -Wall -pedantic -mtune=native -pipe -c vroom_big_int.cc -o vroom_big_int.o
vroom_big_int.cc:23:13: error: use of undeclared identifier 'LONG_LONG_MAX'
  if (val > LONG_LONG_MAX) {
            ^
1 error generated.
make: *** [vroom_big_int.o] Error 1
ERROR: compilation failed for package ‘vroom’
* removing ‘/usr/local/Cellar/r/3.6.2/lib/R/library/vroom’
```

</details>

Any ideas?

**Comment by jimhester** [comment]
I switched to using `LLONG_MAX`, which seems to be the preferred name, see if that fixes your compilation issues. Though I will say unless you have _extremely_ good reasons you will be far better off using `brew cask install r` to install the CRAN distribution of R.

**Comment by rileytwo** [comment]
## Summary

I was able to compile `vroom` successfully, but I think putting this here may be helpful for anyone else who may come along with the same, or similar, problem that I had.

### Install One

The installation failed at first:

```make
/usr/local/bin/ccache /usr/local/opt/llvm/bin/clang++ -std=c++11 -I"/usr/local/Cellar/r/3.6.2/lib/R/include" -DNDEBUG  -I"/usr/local/Cellar/r/3.6.2/lib/R/library/progress/include" -I"/usr/local/Cellar/r/3.6.2/lib/R/library/Rcpp/include" -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -I/usr/local/opt/llvm/include -I/usr/local/opt/openssl/include -I/usr/local/opt/gettext/include -I/usr/local/opt/tcl-tk/include -I/usr/local/opt/jpeg/include -I/usr/local/opt/libpng/include -Imio/include -DWIN32_LEAN_AND_MEAN -Ispdlog/include -fPIC  -g -O3 -Wall -pedantic -mtune=native -pipe -c vroom_big_int.cc -o vroom_big_int.o
   vroom_big_int.cc:3:10: fatal error: 'climits.h' file not found
   #include <climits.h>
            ^~~~~~~~~~~
   1 error generated.
   make: *** [vroom_big_int.o] Error 1
   ERROR: compilation failed for package ‘vroom’
─  removing ‘/private/var/folders/vp/j9vvbjdd60dc1_l6gk470r2h0000gn/T/RtmpeBsc4I/Rinst5383a377231/vroom’
         -----------------------------------
   ERROR: package installation failed
Error: Failed to install 'vroom' from GitHub:
  System command error, exit status: 1, stdout + stderr (last 10 lines):
E> /usr/local/bin/ccache /usr/local/opt/llvm/bin/clang++ -std=c++11 -I"/usr/local/Cellar/r/3.6.2/lib/R/include" -DNDEBUG  -I"/usr/local/Cellar/r/3.6.2/lib/R/library/progress/include" -I"/usr/local/Cellar/r/3.6.2/lib/R/library/Rcpp/include" -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include -I/usr/local/opt/llvm/include -I/usr/local/opt/openssl/include -I/usr/local/opt/gettext/include -I/usr/local/opt/tcl-tk/include -I/usr/local/opt/jpeg/include -I/usr/local/opt/libpng/include -Imio/include -DWIN32_LEAN_AND_MEAN -Ispdlog/include -fPIC  -g -O3 -Wall -pedantic -mtune=native -pipe -c vroom_big_int.cc -o vroom_big_int.o
E> vroom_big_int.cc:3:10: fatal error: 'climits.h' file not found
E> #include <climits.h>
E>          ^~~~~~~~~~~
E> 1 error generated.
E> make: *** [vroom_big_int.o] Error 1
```

From what I can tell, macOS doesn't have a `climits.h` header file (please correct me if I'm wrong).

I found [this](https://www.includehelp.com/cpp-tutorial/LLONG_MAX-constant-with-example.aspx), which led me to believe that `vroom` could be compiled if a `climits.h` existed, even if the `climits.h` was actually a symlink to a `limits.h` file.

### Install Two

There are a few places `limits.h` may be on a mac, but two stood out:

```
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/limits.h
```

and

```
/Library/Developer/CommandLineTools/usr/include/c++/v1/limits.h
```

Knowing just enough about C++ to know that I **don't know anything**, I flipped a coin and symlinked `/Library/Developer/CommandLineTools/usr/include/c++/v1/limits.h` to `/usr/local/include/climits.h`.

I tried installing again, and this time the installation was successful.

### Install Three

But that's kind of messy. I didn't feel all too comfortable just symlinking a file to `/usr/local/include/climits.h`, so I removed the symlink.

I cloned `vroom` and commented out `#include <climits.h>` in [src/vroom_big_int.cc](https://github.com/r-lib/vroom/blob/master/src/vroom_big_int.cc).

`vroom` installed successfully here as well.

### Install Four

While I couldn't find any `climits.h` headers, there are many `climits` files, depending on what someone may have installed; below are just a few that appear when running `mdfind climits`.

```
/usr/local/Cellar/gcc@7/7.5.0_1/include/c++/7.5.0/climits
/usr/local/Cellar/gcc@7/7.5.0_1/include/c++/7.5.0/tr1/climits
/usr/local/Cellar/gcc@8/8.3.0_2/include/c++/8.3.0/climits
/usr/local/Cellar/gcc@8/8.3.0_2/include/c++/8.3.0/tr1/climits
/usr/local/Cellar/llvm/9.0.0_1/Toolchains/LLVM9.0.0.xctoolchain/usr/include/c++/v1/climits
/usr/local/Cellar/llvm/9.0.0_1/include/c++/v1/climits
/usr/local/Cellar/gcc/9.2.0_3/include/c++/9.2.0/climits
/usr/local/Cellar/gcc/9.2.0_3/include/c++/9.2.0/tr1/climits
/usr/local/Cellar/boost/1.72.0/include/boost/compatibility/cpp_c_headers/climits
/Library/Developer/CommandLineTools/usr/include/c++/v1/climits
```

I went back to [src/vroom_big_int.cc](https://github.com/r-lib/vroom/blob/master/src/vroom_big_int.cc), and changed `#include <climits.h>` to `#include <climits>`, and tried installing again.

`vroom` installed successfully, again.

### I like the fourth method the best

But I don't know if changing `<climits.h>` to `<climits>` is how to actually fix my problem, or if this is just slapping on bandaid for my specific use case.

**Comment by rileytwo** [comment]
Fixed in 6cf9ac6. Thanks!
